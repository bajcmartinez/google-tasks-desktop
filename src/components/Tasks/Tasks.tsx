import React, { ChangeEvent, useEffect } from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { Grid } from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import Fab from '@material-ui/core/Fab';
import AddIcon from '@material-ui/icons/Add';
import Zoom from '@material-ui/core/Zoom';
import IconButton from '@material-ui/core/IconButton';
import ListViewIcon from '@material-ui/icons/ViewList';
import ComfortViewIcon from '@material-ui/icons/ViewStream';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import DueDateView from './TaskViews/DueDateView/DueDateView';
import { ISettings } from '../../types';
import TaskEdit from './TaskEdit';
import ListView from './TaskViews/ListView';
import { Task, TaskList as TaskListType } from '../../types/google';

interface IProps {
  settings: ISettings;
  tasks: Task[];
  taskLists: TaskListType[];
  selectedTask?: Task;
  selectedTaskListId: string;
  title: string;
  updateTaskCompletion: (task: string, tasklist: string, completed: boolean) => void;
  setSelectedTask: (task: Task) => void;
  updateTask: (task: Task) => void;
  insertTask: (task: Task) => void;
  deleteTask: (task: Task) => void;
  switchSetting: (key: string, value: any) => void;
}

const useStyles = makeStyles(theme => ({
  section: {
    padding: theme.spacing(1, 2, 3, 2),
    height: 'calc(100vh - 100px)',
    overflowY: 'scroll',
    position: 'relative',
  },

  fab: {
    position: 'absolute',
    bottom: theme.spacing(2),
    right: theme.spacing(2),
  },

  sectionTopBar: {
    display: 'flex',
    alignItems: 'center',
  },

  spacer: {
    flexGrow: 1,
  },
}));

const Tasks: React.FC<IProps> = ({
  settings,
  switchSetting,
  taskLists,
  tasks,
  setSelectedTask,
  selectedTaskListId,
  insertTask,
  updateTask,
  updateTaskCompletion,
  deleteTask,
  selectedTask,
  title,
}) => {
  const classes = useStyles();

  const handleSelectedTaskChanged = (task: Task): void => {
    setSelectedTask(task);
  };

  const handleInsertTask = (): void => {
    insertTask({
      title: '',
      notes: '',
      listId: selectedTaskListId,
      subtasks: [] as Task[],
    } as Task);
  };

  useEffect(() => {
    if (tasks.length > 0)
      if (selectedTask) {
        // Only select the first if the selected task is not anymore on the list
        const st = tasks.find((find: Task) => find.id === selectedTask.id) || tasks[0];
        handleSelectedTaskChanged(st);
      } else {
        handleSelectedTaskChanged(tasks[0]);
      }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tasks, selectedTask]);

  return (
    <div>
      <Grid container spacing={2}>
        <Grid item xs={12} md={6}>
          <Paper className={classes.section}>
            <div className={classes.sectionTopBar}>
              <Typography variant="body1">View:</Typography>
              &nbsp;
              <TextField
                label=""
                value={settings.taskView}
                onChange={(event: ChangeEvent<HTMLInputElement>) => switchSetting('taskView', event.target.value)}
                select
              >
                <MenuItem dense value="DueDateView">
                  Due Date
                </MenuItem>
                <MenuItem dense value="ListView">
                  List
                </MenuItem>
              </TextField>
              <div className={classes.spacer} />
              <IconButton
                color="default"
                aria-label="Switch View"
                size="small"
                onClick={() => switchSetting('comfortView', !settings.comfortView)}
              >
                {settings.comfortView ? (
                  <ComfortViewIcon titleAccess="Switch to Comfort View" />
                ) : (
                  <ListViewIcon titleAccess="Switch to List View" />
                )}
              </IconButton>
            </div>

            {settings.taskView === 'ListView' && (
              <ListView
                settings={settings}
                title={title}
                tasks={tasks}
                handleSelectedTaskChanged={handleSelectedTaskChanged}
                updateTaskCompletion={updateTaskCompletion}
              />
            )}

            {settings.taskView === 'DueDateView' && (
              <DueDateView
                settings={settings}
                tasks={tasks}
                handleSelectedTaskChanged={handleSelectedTaskChanged}
                updateTaskCompletion={updateTaskCompletion}
              />
            )}

            <Zoom in={selectedTaskListId !== 'all'} unmountOnExit>
              <Fab
                aria-label="Insert Task"
                className={classes.fab}
                color="primary"
                data-test-id="tasks-add-button"
                onClick={handleInsertTask}
              >
                <AddIcon />
              </Fab>
            </Zoom>
          </Paper>
        </Grid>

        <Grid item xs={12} md={6}>
          <Paper className={classes.section}>
            <TaskEdit
              task={selectedTask}
              taskLists={taskLists}
              insertTask={insertTask}
              updateTask={updateTask}
              deleteTask={deleteTask}
            />
          </Paper>
        </Grid>
      </Grid>
    </div>
  );
};

export default Tasks;
