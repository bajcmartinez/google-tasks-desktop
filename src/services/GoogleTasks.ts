import moment, { Moment } from 'moment';

// @ts-ignore
let google = window.gapi;

export type TaskList = {
  id: string,
  title: string,
  updatedAt: Moment,
  status: string
}

export type Task = {
  id: string,
  title: string,
  notes?: string,
  completed: boolean,
  completedAt?: Moment,
  dueAt?: Moment,
  parent: string,
  updatedAt: Moment,
  status: string,
  listId: string,
  subtasks: Task[]
}

class GoogleTasksService {
  private clientId: string = "721709625729-0jp536rce8pn3i5ie0pg213d2t55mu55.apps.googleusercontent.com";
  private scopes: string = 'https://www.googleapis.com/auth/tasks';
  private isLoaded: boolean = false;
  private auth: any;

  /**
   * Loads the client library and gets all the api required information from google servers
   *
   */
  load() {
    if (this.isLoaded) return Promise.resolve()

    const self = this;

    return new Promise(resolve => {

      // To load first we need to inject the scripts
      const script = document.createElement("script");
      script.src = "gapi.js";
      // @ts-ignore
      script.onload = () => {
        // @ts-ignore
        google = window.gapi;

        // Then we load the API
        google.load('client', () => {
          google.client.load('tasks', 'v1', () => {
            // @ts-ignore
            google = window.gapi;

            self.isLoaded = true;
            resolve();
          });
        });
      };

      document.body.appendChild(script);
    });
  }

  /**
   * Gets the client authorization to query google's API
   *
   */
  authorize() {
    return new Promise(async (resolve) => {
      await this.load();

      this.auth = await google.auth2.init({
        client_id: this.clientId,
        scope: this.scopes
      });

      resolve();
    });
  }

  /**
   * Returns whether the current session is signed in or not
   *
   */
  isSignedIn () {
    if (!this.auth) return false;
    return this.auth.isSignedIn.get();
  }

  /**
   * Event listener for sign in status
   *
   * @param subscriber
   */
  subscribeSigninStatus (subscriber: (status: boolean) => void) {
    if (!this.auth) return false;
    return this.auth.isSignedIn.listen(subscriber);
  }

  /**
   * Starts the sign in process against your Google Account
   *
   */
  signIn() {
    this.auth.signIn();
  }

  /**
   * Lists all tasks lists
   *
   * @returns TaskList[]
   */
  async listTaskLists() {
    const response = await google.client.tasks.tasklists.list();

    return response.result.items.map((item: any): TaskList => ({
      id: item["id"],
      title: item["title"],
      updatedAt: moment(item["updated"])
    }) as TaskList);
  }

  /**
   * Lists all tasks for a given task list
   *
   * @returns Task[]
   */
  async listTasks(taskListId: string) {
    const response = await google.client.tasks.tasks.list({
      tasklist: taskListId,
      showCompleted: true,
      showHidden: true
    });

    const items = response.result.items;

    const mapItems = (tasks: any[]): Task[] => {
      return tasks.map((item: any): Task => ({
        id: item["id"],
        title: item["title"],
        notes: item["notes"],
        dueAt: item["due"] ? moment(item["due"]) : undefined,
        parent: item["parent"],
        completed: item["status"] === "completed",
        completedAt: item["completed"] ? moment(item["completed"]) : undefined,
        updatedAt: moment(item["updated"]),
        listId: taskListId,
        status: item["status"],
        subtasks: mapItems(items.filter((subitem: any) => subitem["parent"] === item["id"]))
      }) as Task);
    };

    return mapItems(items);
  }

  async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    await google.client.tasks.tasks.update({
      tasklist,
      task,
      id: task,
      status: completed ? 'completed' : 'needsAction'
    });
  }
}

export default new GoogleTasksService();