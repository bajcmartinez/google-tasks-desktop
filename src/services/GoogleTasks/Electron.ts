/* istanbul ignore file */
import { GoogleTasksWebService } from './Web';
import { google, tasks_v1 } from 'googleapis';
import moment  from 'moment';
import { Task, TaskList, OAuthKeys } from '../../types/google';
import { APIEndpoint, OAuth2Client } from 'googleapis-common';
import { ipcRenderer, remote } from 'electron';
import * as fs from 'fs';

const scopes: Array<string> = ['https://www.googleapis.com/auth/tasks'];
const tokenStorageId = "gatt-v1";
const credentialsFileName = 'google-tasks-ui.json'

export class GoogleTasksElectronService extends GoogleTasksWebService {
  private tasksAPI: APIEndpoint | undefined;
  private taskListsAPI: APIEndpoint | undefined;
  private oAuth2Client: OAuth2Client | undefined;
  private signedInCallback: (isSignedIn: boolean) => void = () => {};

  getCredentials(): OAuthKeys | undefined {
    const possible = [
      remote.app.getPath('home') + '/' + credentialsFileName,
      remote.app.getPath('home') + '/.' + credentialsFileName,
      remote.app.getPath('documents') + '/' + credentialsFileName,
      remote.app.getPath('documents') + '/.' + credentialsFileName,
      remote.app.getPath('desktop') + '/' + credentialsFileName,
      remote.app.getPath('desktop') + '/.' + credentialsFileName
    ]

    let credentials: OAuthKeys | undefined = undefined;
    possible.some((filename) => {
      if (fs.existsSync(filename)) {
        const data = fs.readFileSync(filename, 'utf8');
        credentials = JSON.parse(data);
        return true;
      }

      return false;
    });

    return credentials;
  }

  /**
   * Gets the client authorization to query google's API
   *
   */
  load(callback: (isSignedIn: boolean) => void) {
    // Look up for the keys
    const credentials = this.getCredentials();

    if (!credentials)
      return Promise.reject("Missing credentials file!");

    this.oAuth2Client = new google.auth.OAuth2(
      credentials.installed.client_id,
      credentials.installed.client_secret,
      credentials.installed.redirect_uris[0]
    );

    this.oAuth2Client.on('tokens', (tokens) => {
      this.oAuth2Client && this.oAuth2Client.setCredentials(tokens);
      localStorage.setItem(tokenStorageId, JSON.stringify(tokens));
    });

    const { tasks: tasksAPI, tasklists: taskListsAPI } = google.tasks({
      version: 'v1',
      auth: this.oAuth2Client
    });

    this.tasksAPI = tasksAPI;
    this.taskListsAPI = taskListsAPI;
    this.signedInCallback = callback;

    // Check if we are signed in
    if (this.isSignedIn()) {
      this.signIn();
    }

    ipcRenderer.on('google-auth-reply', async (event:any, access_token: string) => {
      await this.getTokenAPI(access_token);
      if (this.signedInCallback)
        this.signedInCallback(true);
    });

    return Promise.resolve();
  }

  async getTokenAPI(code: string) {
    if (this.oAuth2Client) {
      try {
        const { tokens } = await this.oAuth2Client.getToken(code);
        this.oAuth2Client.setCredentials(tokens);
        localStorage.setItem(tokenStorageId, JSON.stringify(tokens));
        return tokens;
      } catch (err) {
        return Promise.reject(err);
      }
    } else {
      return Promise.reject();
    }
  }

  /**
   * Returns whether the current session is signed in or not
   *
   */
  isSignedIn () {
    return !!localStorage.getItem(tokenStorageId);
  }

  /**
   * Starts the sign in process against your Google Account
   *
   */
  signIn(consent: boolean = false) {
    if (this.oAuth2Client) {
      const authorizeUrl = this.oAuth2Client.generateAuthUrl({
        access_type: 'offline',
        scope: scopes,
        prompt: consent ? 'consent' : undefined
      });

      const token = localStorage.getItem(tokenStorageId);
      if (!token) {
        ipcRenderer.send('google-auth-start', authorizeUrl);
        return Promise.resolve();
      } else {
        this.oAuth2Client.setCredentials(JSON.parse(token));
        this.signedInCallback(true);
        return Promise.resolve();
      }
    }

    return Promise.reject();
  }

  /**
   * Starts the sign out process against your Google Account
   *
   */
  signOut() {
    localStorage.removeItem(tokenStorageId);
    this.signedInCallback(false);
  }

  /**
   * Lists all tasks lists
   *
   * @returns TaskList[]
   */
  async listTaskLists() {
    const response = this.taskListsAPI && await this.taskListsAPI.list();

    if (!response.data || !response.data.items)
      return [];

    return response.data.items.map((item: any): TaskList => ({
      id: item["id"],
      title: item["title"],
      updatedAt: moment(item["updated"])
    }) as TaskList);
  }

  /**
   * Lists all tasks for a given task list
   *
   * @returns Task[]
   */
  async listTasks(taskListId: string, pageToken: string = '') {
    const response = this.tasksAPI && await this.tasksAPI.list({
      tasklist: taskListId,
      showCompleted: false,
      showHidden: true,
      pageToken: pageToken
    });

    let result: Task[] = [];


    if (!response.data || !response.data.items)
      return result;

    const items = response.data.items;

    const nextPageToken = response.data.nextPageToken as string;

    const mapItems = (tasks: any[]): Task[] => {
      return tasks.map((item: any): Task => ({
        id: item["id"],
        title: item["title"] ? item['title'] : '',
        notes: item["notes"] ? item["notes"] : '',
        dueAt: item["due"] ? moment(item["due"]) : undefined,
        parent: item["parent"],
        completed: item["status"] === "completed",
        completedAt: item["completed"] ? moment(item["completed"]) : undefined,
        updatedAt: moment(item["updated"]),
        listId: taskListId,
        status: item["status"],
        isDirty: false,
        subtasks: mapItems(items.filter((subitem: any) => subitem["parent"] === item["id"]))
      }) as Task);
    };

    result = result.concat(mapItems(items.filter((subitem: any) => !subitem["parent"])));

    if (nextPageToken) {
      result = result.concat(await this.listTasks(taskListId, nextPageToken));
    }

    return result;
  }

  /**
   * Sets the completion status of a task
   *
   * @param task: string
   * @param tasklist: string
   * @param completed: boolean
   */
  async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    const requestBody: tasks_v1.Schema$Task = {
      id: task,
      status: completed ? 'completed' : 'needsAction'
    }

    this.tasksAPI && await this.tasksAPI.update({
      tasklist,
      task,
      requestBody
    });
  }

  /**
   * Creates a task
   *
   * @param task: Task
   */
  async insertTask(task: Task) {
    return await google.client.tasks.tasks.insert({
      tasklist: task.listId,
      task: task.id,
      id: task.id,
      title: task.title ? task.title : '',
      notes: task.notes ? task.notes : '',
      due: task.dueAt ? task.dueAt.format() : null,
      status: task.completed ? 'completed' : 'needsAction',
      parent: task.parent
    });
  }

  /**
   * Updates a task
   *
   * @param task: Task
   */
  async updateTask(task: Task) {
    await google.client.tasks.tasks.update({
      tasklist: task.listId,
      task: task.id,
      id: task.id,
      title: task.title,
      notes: task.notes,
      due: task.dueAt ? task.dueAt.format() : null,
      status: task.completed ? 'completed' : 'needsAction'
    });
  }

  /**
   * Deletes a task
   *
   * @param task: string
   * @param tasklist: string
   */
  async deleteTask(task: string, tasklist: string) {
    await google.client.tasks.tasks.delete({
      tasklist,
      task
    });
  }

  /**
   * Only used for testing purposes, view mock file
   */
  reset() {

  }
}