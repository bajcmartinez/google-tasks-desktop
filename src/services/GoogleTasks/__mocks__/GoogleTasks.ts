class GoogleTasksService {

  isSignedIn() {
    return Promise.resolve(true);
  }

  listTaskLists() {
    return Promise.resolve([]);
  }
}

export default new GoogleTasksService();