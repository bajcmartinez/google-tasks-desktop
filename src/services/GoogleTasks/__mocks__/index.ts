import moment from 'moment';
import { Task } from "../../../types/google";
import { GoogleTasksWebService } from '../Web';

class Index extends GoogleTasksWebService {

  private _signedIn: boolean = false;
  private _subscription: ((status: boolean) => void) | undefined;

  load(callback: (isSignedIn: boolean) => void) {
    this.subscribeSigninStatus(callback);
    return Promise.resolve();
  }

  authorize() {
    return Promise.resolve(true);
  }

  signIn() {
    this._signedIn = true;
    this._subscription && this._subscription(this._signedIn);
  }

  signOut() {
    this._signedIn = false;
    this._subscription && this._subscription(this._signedIn);
  }

  isSignedIn() {
    return this._signedIn;
  }

  listTaskLists() {
    return Promise.resolve([
      {
        id: "1",
        status: "active",
        title: "Task List 1",
        updatedAt: moment()
      },
      {
        id: "2",
        status: "active",
        title: "Task List 2",
        updatedAt: moment()
      }
    ]);
  }

  async listTasks(taskListId: string, pageToken: string = '') {
    return Promise.resolve([
      {
        id: `${taskListId.toString()}_1`,
        title: 'Task 1',
        notes: 'notes 1',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(7, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_2`,
        title: 'Task 2',
        notes: 'notes 2',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_3`,
        title: 'Task 3',
        notes: 'notes 3',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(-3, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_4`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_5`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "completed",
        updatedAt: moment(),
        parent: `${taskListId.toString()}_4`
      }
    ]);
  }

  async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    return Promise.resolve();
  }

  async insertTask(task: Task) {
    return Promise.resolve({
      result: {
        id: 999
      }
    });
  }

  async deleteTask(task: string, tasklist: string) {
    return Promise.resolve();
  }

  async updateTask(task: Task) {
    return Promise.resolve();
  }

  subscribeSigninStatus (subscriber: (status: boolean) => void) {
    this._subscription = subscriber;
  }

  reset() {
    this._signedIn = false;
    // @ts-ignore
    this._subscription = null;
  }
}

const r = new Index();
export default r;