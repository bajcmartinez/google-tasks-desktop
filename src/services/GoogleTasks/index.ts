/* istanbul ignore file */
import { GoogleTasksWebService } from './Web'

let GoogleTasksService = GoogleTasksWebService;

if (typeof process !== 'undefined' && process.versions && process.versions.hasOwnProperty('electron')) {
  console.log('here I am');
  const { GoogleTasksElectronService } = require('./Electron');
  GoogleTasksService = GoogleTasksElectronService;
}

const r = new GoogleTasksService();
export default r;