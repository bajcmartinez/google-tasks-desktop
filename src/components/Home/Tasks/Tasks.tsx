import React, { useEffect, useState } from 'react'
import makeStyles from '@material-ui/core/styles/makeStyles';
import { Task, TaskList as TaskListType } from '../../../services/GoogleTasks';
import { Grid } from '@material-ui/core';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import TaskList from './TaskList';
import TaskEdit from './TaskEdit'

interface IProps {
  tasks: Task[],
  taskLists: TaskListType[],
  title: string,
  updateTaskCompletion: (task: string, tasklist: string, completed: boolean) => void,
  updateTask: (task: Task) => void
  deleteTask: (task: Task) => void
}

const useStyles = makeStyles(theme => ({
  section: {
    padding: theme.spacing(3, 2),
    height: 'calc(100vh - 176px)',
    overflowY: 'scroll'
  },

  fixedSection: {
    padding: theme.spacing(3, 2),
    position: 'fixed',
    height: 'calc(100vh - 176px)',
  }
}));

const Tasks: React.FC<IProps> = (props) => {
  const classes = useStyles();

  const [selectedTask, setSelectedTask] = useState<Task>();

  function handleSelectedTaskChanged(task: Task):void {
    setSelectedTask(task);
  }

  const [ tasks, setTasks ] = useState<Task[]>([])
  useEffect(() => {
    setTasks(props.tasks.sort((a: Task, b: Task) => ((a.dueAt || new Date()) < (b.dueAt || new Date())) ? -1 : 1));
  }, [props.tasks]);

  useEffect(() => {
    if (tasks.length > 0)
      if (selectedTask) {
        // Only select the first if the selected task is not anymore on the list
        const st = tasks.find((find: Task) => find.id === selectedTask.id) || tasks[0];
        handleSelectedTaskChanged(st);
      } else {
        handleSelectedTaskChanged(tasks[0]);
      }
  }, [tasks, selectedTask]);


  return (
    <div>
      <Grid container spacing={2}>
        <Grid item xs={12} md={6}>
          <Paper className={classes.section}>
            <Typography variant="h6">
              {props.title}
            </Typography>

            <TaskList tasks={tasks} handleSelectedTaskChanged={handleSelectedTaskChanged} updateTaskCompletion={props.updateTaskCompletion} />
          </Paper>
        </Grid>

        <Grid item xs={12} md={6}>
          <Paper className={classes.section}>
            <TaskEdit
                task={selectedTask}
                taskLists={props.taskLists}
                updateTask={props.updateTask}
                deleteTask={props.deleteTask}
            />
          </Paper>
        </Grid>
      </Grid>


    </div>
  );
};

export default Tasks;
