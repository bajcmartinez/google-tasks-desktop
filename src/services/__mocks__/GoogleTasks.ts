import moment from 'moment';
import {Task} from "../GoogleTasks";

class GoogleTasksService {

  static _signedIn: boolean = false;
  static _subscription: (status: boolean) => void | undefined;

  static authorize() {
    return Promise.resolve(true);
  }

  static signIn() {
    this._signedIn = true;
    this._subscription && this._subscription(this._signedIn);
  }

  static signOut() {
    this._signedIn = false;
    this._subscription && this._subscription(this._signedIn);
  }

  static isSignedIn() {
    return this._signedIn;
  }

  static listTaskLists() {
    return Promise.resolve([
      {
        id: "1",
        status: "active",
        title: "Task List 1",
        updatedAt: moment()
      },
      {
        id: "2",
        status: "active",
        title: "Task List 2",
        updatedAt: moment()
      }
    ]);
  }

  static async listTasks(taskListId: string, pageToken: string = '') {
    return Promise.resolve([
      {
        id: `${taskListId.toString()}_1`,
        title: 'Task 1',
        notes: 'notes 1',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(7, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_2`,
        title: 'Task 2',
        notes: 'notes 2',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_3`,
        title: 'Task 3',
        notes: 'notes 3',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(-3, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_4`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_5`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "completed",
        updatedAt: moment(),
        parent: `${taskListId.toString()}_4`
      }
    ]);
  }

  static async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    return Promise.resolve();
  }

  static async insertTask(task: Task) {
    return Promise.resolve({
      result: {
        id: 999
      }
    });
  }

  static async deleteTask(task: string, tasklist: string) {
    return Promise.resolve();
  }

  static async updateTask(task: Task) {
    return Promise.resolve();
  }

  static subscribeSigninStatus (subscriber: (status: boolean) => void) {
    this._subscription = subscriber;
  }

  /**
   * Only used for testing purposes
   */
  static reset() {
    this._signedIn = false;
    // @ts-ignore
    this._subscription = null;
  }
}

export default GoogleTasksService;