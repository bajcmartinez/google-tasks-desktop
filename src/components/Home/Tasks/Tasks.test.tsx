import React from 'react';
import { mount, shallow } from 'enzyme'
import Tasks from './Tasks';
import { Task, TaskList } from '../../../services/GoogleTasks'
import moment from 'moment';
import MomentUtils from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { Fab } from '@material-ui/core'


describe("Basic", () => {
  it('should render without crashing', () => {
    const taskList: TaskList = {
      id: "1",
      status: "active",
      title: "Task List 1",
      updatedAt: moment()
    }

    const titleBar = shallow(
      <Tasks
        tasks={[]}
        selectedTaskListId="1"
        taskLists={[taskList]}
        title="Task List 1"
        insertTask={() => null}
        updateTask={() => null}
        updateTaskCompletion={() => null}
        deleteTask={() => null}
        setSelectedTask={() => null}
      />
    );
    expect(titleBar.exists()).toBe(true);
  });
});

describe("Events", () => {

  it('should trigger the selected item event', () => {
    const handler = jest.fn();

    const taskList: TaskList = {
      id: "1",
      status: "active",
      title: "Task List 1",
      updatedAt: moment()
    }

    const tasks: Task[] = [
      {
        id: '1',
        title: 'Task 1',
        notes: 'notes',
        subtasks: [],
        listId: '1',
        isDirty: false,
        dueAt: moment(),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: '2',
        title: 'Task 2',
        notes: 'notes 2',
        subtasks: [],
        listId: '1',
        isDirty: false,
        dueAt: moment(),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      }
    ]

    mount(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <Tasks
          tasks={tasks}
          selectedTaskListId="1"
          taskLists={[taskList]}
          title="Task List 1"
          insertTask={() => null}
          updateTask={() => null}
          updateTaskCompletion={() => null}
          deleteTask={() => null}
          setSelectedTask={handler}
        />
      </MuiPickersUtilsProvider>
    );

    expect(handler).toBeCalled();
  });

  it('should trigger the insert item event', () => {
    const handler = jest.fn();

    const taskList: TaskList = {
      id: "1",
      status: "active",
      title: "Task List 1",
      updatedAt: moment()
    }

    const tasks: Task[] = [
      {
        id: '1',
        title: 'Task 1',
        notes: 'notes',
        subtasks: [],
        listId: '1',
        isDirty: false,
        dueAt: moment(),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      }
    ]

    const wrapper = mount(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <Tasks
          tasks={tasks}
          selectedTaskListId="1"
          taskLists={[taskList]}
          title="Task List 1"
          insertTask={handler}
          updateTask={() => null}
          updateTaskCompletion={() => null}
          deleteTask={() => null}
          setSelectedTask={() => null}
        />
      </MuiPickersUtilsProvider>
    );

    const userButton = wrapper.find(Fab).first();
    userButton.simulate('click');

    expect(handler).toBeCalled();
  });

});
