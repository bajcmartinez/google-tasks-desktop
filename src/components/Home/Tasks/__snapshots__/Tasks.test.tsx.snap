// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic should render a <Tasks /> component as expected 1`] = `ReactWrapper {}`;

exports[`Basic should render a <Tasks /> component as expected with a selected item 1`] = `ReactWrapper {}`;
